src/Search/Services/SearchableService.php

```diff
@@ -0,0 +1,162 @@
+
+/**
+ * Checks if a DataObject is publicly viewable, thus able to be added to or retrieved from a publicly searchable index.
+ * Results are cached because these checks may be run multiple times, as there a few different code paths that search
+ * results might follow in real-world search implementations.
+ */
+class SearchableService
+{
+    use Injectable;
+    use Extensible;
+    use Configurable;
+
+    /**
+     * Skip the canView() check at a class level to increase performance of search reindex.
+     * Be careful as this may lead to content showing in search results that should not be there such as non-public,
+     * cms-user-only content.  This may potentially happen via edge cases such as skipping checks where subclasses
+     * are involved.
+     *
+     * This has no effect on when search results as canView() must still be run there
+     *
+     * @var array namespaced classes to skip canView() check on search reindex
+     */
+    private static $classes_that_skip_indexing_canview_check = [];
+
+    /**
+     * Non-persistant memory cache that only lasts the lifetime of the request
+     *
+     * @var array
+     */
+    private $cache = [];
+
+    /**
+     * Clears the internal cache
+     */
+    public function clearCache(): void
+    {
+        $this->cache = [];
+    }
+
+    /**
+     * Used during search reindex
+     *
+     * This is considered the primary layer of protection
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    public function isIndexable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, true);
+    }
+
+    /**
+     * Used when retrieving search results
+     *
+     * This is considered the secondary layer of protection
+     *
+     * It's important to still have this layer in conjuction with the index layer as non-searchable results may be
+     * in the search index because:
+     * a) they were added to the index pre-fulltextsearch 3.7 and a reindex to purge old records was never run, OR
+     * b) the DataObject has a non-deterministic canView() check such as `return $date <= $dateOfIndex;`
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    public function isViewable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, false);
+    }
+
+    /**
+     * Checks and caches whether the given DataObject can be indexed. This is determined by two factors:
+     * - Whether the ShowInSearch property / getShowInSearch() method evaluates to true
+     * - Whether the canView method evaluates to true against an anonymous user (optional, can be disabled)
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    private function isSearchable(DataObject $obj, bool $indexing): bool
+    {
+        // check if is a valid DataObject that has been persisted to the database
+        if (is_null($obj) || !$obj->ID) {
+            return false;
+        }
+
+        $key = $this->getCacheKey($obj);
```

Comment:
Have added it.  Technically wasn't required because the cache is only held in memory and is non-persistent, and you're not going to do an index and search retrieving in the same request.  Still, no harm in added it and it may be helpful for anyone using the extension hooks